pkg/plugins/gateway/algorithms/vtc/token_tracker.go

```diff
@@ -119,79 +155,224 @@ func (t *InMemorySlidingWindowTokenTracker) getCutoffTimestamp() int64 {
+			total = t.userTotals[user]
+		}
 		t.mu.Unlock()
 	}
 
-	return sum, nil
+	return total, nil
 }
 
+// Time: O(1) | Space: O(1)
+func (t *InMemorySlidingWindowTokenTracker) GetMinTokenCount(ctx context.Context) (float64, error) {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+
+	// Return default min if no active users (minTrackedToken is still at initialization value)
+	if t.minTrackedToken == math.MaxFloat64 {
+		return tokenTrackerMinTokens, nil
+	}
+	return t.minTrackedToken, nil
+}
+
+// Time: O(1) | Space: O(1)
+func (t *InMemorySlidingWindowTokenTracker) GetMaxTokenCount(ctx context.Context) (float64, error) {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+
+	// If no active users or all have zero tokens, return default max
+	if t.maxTrackedToken == 0 {
+		return tokenTrackerMaxTokens, nil
+	}
+	return t.maxTrackedToken, nil
+}
+
+// Time: Avg O(1) (amortized), Worst O(B_u) where B_u = buckets for user u | Space: O(1)
 func (t *InMemorySlidingWindowTokenTracker) UpdateTokenCount(ctx context.Context, user string, inputTokens, outputTokens float64) error {
+	if user == "" {
+		return fmt.Errorf("user ID cannot be empty")
+	}
+
 	t.mu.Lock()
 	defer t.mu.Unlock()
 
-	if user == "" {
-		return nil
-	}
 	now := time.Now()
+	currentTimestamp := t.bucketUnit.toTimestamp(now)
+	cutoff := t.getCutoffTimestamp()
 
-	truncatedTime := t.bucketUnit.truncateTime(now)
-	windowStart := t.bucketUnit.toTimestamp(truncatedTime)
-
-	tokens := inputTokens*t.config.InputTokenWeight + outputTokens*t.config.OutputTokenWeight
-	buckets, ok := t.userBuckets[user]
+	_, ok := t.userBucketStore[user]
 	if !ok {
-		buckets = make(map[int64]float64)
-		t.userBuckets[user] = buckets
+		t.userBucketStore[user] = &userBucketData{
+			buckets: list.New(),
+			lookup:  make(map[int64]*list.Element),
+		}
 	}
-	buckets[windowStart] += tokens
 
-	cutoff := t.getCutoffTimestamp()
-	t.pruneExpiredBuckets(user, cutoff)
+	oldTotal := t.userTotals[user]
+
+	// Prune first before adding/updating to maintain window size constraint accurately
+	t.pruneExpiredBucketsAndUpdateState(user, cutoff)
+
+	totalAfterPruning := t.userTotals[user]
+	newTokens := inputTokens*t.config.InputTokenWeight + outputTokens*t.config.OutputTokenWeight
+
+	// Check if a bucket for the current timestamp already exists
+	if element, exists := t.userBucketStore[user].lookup[currentTimestamp]; exists {
+		node := element.Value.(*bucketNode)
+		node.tokenCount += newTokens
+	} else {
+		node := &bucketNode{timestamp: currentTimestamp, tokenCount: newTokens}
+		element := t.userBucketStore[user].buckets.PushBack(node)
+		t.userBucketStore[user].lookup[currentTimestamp] = element
+	}
+
+	// Calculate the final new total
+	newTotal := totalAfterPruning + newTokens
+
+	// Update user total and global min/max efficiently
+	t.updateUserTotalAndMinMax(user, oldTotal, newTotal)
 
 	return nil
 }
+
+func (t *InMemorySlidingWindowTokenTracker) updateUserTotalAndMinMax(user string, oldTotal, newTotal float64) {
```

Comment:
I see some logics like recalcMin relies on some status like `len(t.totalsToUsers[oldTotal]) == 0`. I am not that sure whether lock is needed here

Reply:
added comment that caller needs to hold lock